// @evolu/common/dist/src/Console.js
/** Creates a {@link Console} for logging with configurable output. */
export const createConsole = (config = {}) => {
    const instance = {
        enabled: config.enableLogging ?? false,
        log: (...args) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.log(...args);
        },
        info: (...args) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.info(...args);
        },
        warn: (...args) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.warn(...args);
        },
        error: (...args) => {
            // Always log errors, even if disabled
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            console.error(...args);
        },
        debug: (...args) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.debug(...args);
        },
        time: (label) => {
            // eslint-disable-next-line no-console
            if (instance.enabled)
                console.time(label);
        },
        timeLog: (label, ...data) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.timeLog(label, ...data);
        },
        timeEnd: (label) => {
            // eslint-disable-next-line no-console
            if (instance.enabled)
                console.timeEnd(label);
        },
        dir: (object, options) => {
            // eslint-disable-next-line no-console
            if (instance.enabled)
                console.dir(object, options);
        },
        table: (tabularData, properties) => {
            // eslint-disable-next-line no-console
            if (instance.enabled)
                console.table(tabularData, properties);
        },
        count: (label) => {
            // eslint-disable-next-line no-console
            if (instance.enabled)
                console.count(label);
        },
        countReset: (label) => {
            // eslint-disable-next-line no-console
            if (instance.enabled)
                console.countReset(label);
        },
        assert: (value, message, ...optionalParams) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.assert(value, message, ...optionalParams);
        },
        trace: (message, ...optionalParams) => {
            // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
            if (instance.enabled)
                console.trace(message, ...optionalParams);
        },
    };
    return instance;
};
/** Creates a console instance with timestamp prefixes. */
export const createConsoleWithTime = (config = { timestampType: "relative" }) => {
    const console = createConsole(config);
    const startTime = performance.now();
    const getTimestamp = () => {
        if (config.timestampType === "relative") {
            const elapsed = (performance.now() - startTime) / 1000;
            // Format for better readability at different time scales
            if (elapsed < 60) {
                // Under 1 minute: show seconds with millisecond precision
                return `+${elapsed.toFixed(3)}s`;
            }
            else if (elapsed < 3600) {
                // 1 minute to 1 hour: show minutes and seconds with millisecond precision
                const minutes = Math.floor(elapsed / 60);
                const seconds = (elapsed % 60).toFixed(3);
                return `+${minutes}m${seconds}s`;
            }
            else {
                // Over 1 hour: show hours, minutes, and seconds with millisecond precision
                const hours = Math.floor(elapsed / 3600);
                const minutes = Math.floor((elapsed % 3600) / 60);
                const seconds = ((elapsed % 3600) % 60).toFixed(3);
                return `+${hours}h${minutes}m${seconds}s`;
            }
        }
        else {
            // Absolute time - format as HH:MM:SS.mmm
            const now = new Date();
            const hours = now.getHours().toString().padStart(2, "0");
            const minutes = now.getMinutes().toString().padStart(2, "0");
            const seconds = now.getSeconds().toString().padStart(2, "0");
            const milliseconds = now.getMilliseconds().toString().padStart(3, "0");
            return `${hours}:${minutes}:${seconds}.${milliseconds}`;
        }
    };
    const withTimestamp = (fn) => (...args) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        fn(`[${getTimestamp()}]`, ...args);
    };
    // Override methods that should have timestamps
    console.log = withTimestamp(console.log);
    console.info = withTimestamp(console.info);
    console.warn = withTimestamp(console.warn);
    console.error = withTimestamp(console.error);
    console.debug = withTimestamp(console.debug);
    console.trace = withTimestamp(console.trace);
    return console;
};
