// @evolu/common/dist/src/Order.js
/**
 * Creates an ordering function from a "less than" comparator.
 *
 * ### Example
 *
 * ```ts
 * const orderNumber = createOrder<number>((x, y) => x < y);
 * expect(orderNumber(1, 2)).toEqual(-1);
 * expect(orderNumber(2, 1)).toEqual(1);
 * expect(orderNumber(1, 1)).toEqual(0);
 * ```
 */
export const createOrder = (isLessThan) => (x, y) => x === y ? 0 : isLessThan(x, y) ? -1 : 1;
/**
 * Returns an order that reverses the order of the given order.
 *
 * ### Example
 *
 * ```ts
 * reverseOrder(orderNumber)(1, 2); // 1
 * reverseOrder(orderNumber)(2, 1); // -1
 * reverseOrder(orderNumber)(1, 1); // 0
 * ```
 */
export const reverseOrder = (order) => (a, b) => order(b, a);
/**
 * An order for `string` values in ascending order.
 *
 * ### Example
 *
 * ```ts
 * orderString("a", "b"); // -1
 * orderString("b", "a"); // 1
 * orderString("a", "a"); // 0
 * ["c", "b", "a"].toSorted(orderString); // ["a", "b", "c"]
 * ```
 */
export const orderString = createOrder((a, b) => a < b);
/**
 * An order for numbers in ascending order.
 *
 * ### Example
 *
 * ```ts
 * orderNumber(1, 2); // -1
 * orderNumber(2, 1); // 1
 * orderNumber(1, 1); // 0
 * [2, 1, 3].toSorted(orderNumber); // [1, 2, 3]
 * reverseOrder(orderNumber)(1, 2); // 1
 * reverseOrder(orderNumber)(2, 1); // -1
 * reverseOrder(orderNumber)(1, 1); // 0
 * ```
 */
export const orderNumber = createOrder((a, b) => a < b);
/**
 * An order for bigints in ascending order.
 *
 * ### Example
 *
 * ```ts
 * orderBigInt(1n, 2n); // -1
 * orderBigInt(2n, 1n); // 1
 * orderBigInt(1n, 1n); // 0
 * [2n, 1n, 3n].toSorted(orderBigInt); // [1n, 2n, 3n]
 * ```
 */
export const orderBigInt = createOrder((a, b) => a < b);
/** An {@link Order} for Uint8Array. */
export const orderUint8Array = (a, b) => {
    if (a.byteLength > b.byteLength)
        return 1;
    if (a.byteLength < b.byteLength)
        return -1;
    for (let i = 0; i < a.byteLength; i++) {
        if (a[i] < b[i])
            return -1;
        if (a[i] > b[i])
            return 1;
    }
    return 0;
};
